pkg/operator/controller/controller_dns_daemonset.go

```diff
@@ -158,13 +153,88 @@ func (r *reconciler) updateDNSDaemonSet(current, desired *appsv1.DaemonSet) (boo
 
 	// Diff before updating because the client may mutate the object.
 	diff := cmp.Diff(current, updated, cmpopts.EquateEmpty())
+	if safe, reason, err := r.daemonsetUpdateIsSafe(current, updated); err != nil {
+		return false, err
+	} else if !safe {
+		logrus.Warnf("skipping unsafe update to dns daemonset %s/%s: %s; update: %v", updated.Namespace, updated.Name, reason, diff)
+		return false, nil
+	}
+
 	if err := r.client.Update(context.TODO(), updated); err != nil {
 		return false, fmt.Errorf("failed to update dns daemonset %s/%s: %v", updated.Namespace, updated.Name, err)
 	}
 	logrus.Infof("updated dns daemonset %s/%s: %v", updated.Namespace, updated.Name, diff)
 	return true, nil
 }
 
+// daemonsetUpdateIsSafe takes current and updated daemonsets, checks if the
+// update is safe, and returns a Boolean value and a string value indicating
+// whether the update is safe or the reason why it is not.
+func (r *reconciler) daemonsetUpdateIsSafe(current, updated *appsv1.DaemonSet) (bool, string, error) {
+	// Allow the update if the current daemonset doesn't even have a
+	// valid selector.
+	selector, err := metav1.LabelSelectorAsSelector(current.Spec.Selector)
+	if err != nil {
+		return true, "", nil
+	}
+	listOpts := []client.ListOption{
+		client.MatchingLabelsSelector{Selector: selector},
+		client.InNamespace(current.Namespace),
+	}
+	podList := &corev1.PodList{}
+	if err := r.cache.List(context.TODO(), podList, listOpts...); err != nil {
+		return false, "", fmt.Errorf("failed to list the daemonset's pods: %w", err)
+	}
+	// Allow the update if the current daemonset has 0 pods as the update
+	// cannot reduce the number of pods below 0.
+	if len(podList.Items) == 0 {
+		return true, "", nil
+	}
+	nodeLabelSelector := &metav1.LabelSelector{
+		MatchLabels: updated.Spec.Template.Spec.NodeSelector,
+	}
+	nodeSelector, err := metav1.LabelSelectorAsSelector(nodeLabelSelector)
+	if err != nil {
+		return true, "", nil
+	}
+	listOpts = []client.ListOption{
+		client.MatchingLabelsSelector{Selector: nodeSelector},
+	}
+	nodeList := &corev1.NodeList{}
+	if err := r.cache.List(context.TODO(), nodeList, listOpts...); err != nil {
+		return false, "", fmt.Errorf("failed to list nodes: %w", err)
+	}
+	for _, node := range nodeList.Items {
+		if node.Spec.Unschedulable {
+			continue
+		}
+		taints := node.Spec.Taints
+		tolerations := updated.Spec.Template.Spec.Tolerations
+		if !tolerationsTolerateTaints(tolerations, taints) {
+			continue
+		}
+		// We have found a node to which it should be possible to
+		// schedule pods for the updated daemonset, so the update
+		// can be considered safe.
+		return true, "", nil
+	}
+	return false, "updated daemonset's node placement parameters match 0 nodes", nil
```

Comment:
Maybe we should clarify this error message a bit more. How about the following?
```suggestion
	return false, "updated daemonset's node placement parameters match 0 schedulable nodes", nil
```

Reply:
Done.  Thanks!